src/optical_spi.c

```diff
@@ -2,143 +2,139 @@
-*/
-void send_opt_spi_cmd(uint8_t field_num) {
-    // print("Sending optical command - field #%u\n", field_num);
+
+// sends 2 bytes to request reading
+// byte 1 = cmd_byte
+// byte 2 = well data
+void send_opt_spi_cmd(uint8_t cmd_opcode, uint8_t well_data) {
+    // print("Sending optical the command code %X\n", cmd_opcode);
 
     opt_spi_data = 0;
-    opt_spi_num_bytes = 0;
+    spi_frame_number = 0;
 
-    // Send the command to PAY-Optical to start reading data
+    // Send the command to PAY-Optical to start reading data    
     set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
-    uint8_t spi_tx = (0b11 << 6) | field_num;
-    send_spi(spi_tx);
+    send_spi(cmd_opcode);
     set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
 
-    // After this, PAY-Optical should set DATA_RDY high, which will trigger the
-    // interrupt below to receive the data
+    // wait for DATA_RDYn to go LOW, signaling optical is done reading byte 1
+    // --> aka loop until DATA_RDYn is no longer HIGH
+    while (get_data_pin() == 1);
+
+    // print("Sending optical the well data %X\n", well_data);
+    set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
+    send_spi(well_data);
+    set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
 }
 
 
-/*
-This interrupt is triggered when the DATA_RDY pin goes set_cs_high
-(set by PAY-Optical to indicate data is ready).
-The SPI transfer should already be in progress if we get this interrupt.
-*/
-ISR(INT1_vect) {
-    // print("INT1 - DATA_RDY\n");
+void receive_opt_data(uint8_t num_expected_bytes){
+    
+    // still expecting more data from SSM
+    while (num_expected_bytes > spi_frame_number){
 
-    // Get the next byte over SPI
-    set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
-    uint8_t new_byte = send_spi(0x00);
-    set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
+        // OPTICAL ready to send data
+        if (get_data_pin() == 1){
+            uint8_t new_opt_data = SPDR;
+            opt_spi_data = opt_spi_data<<8 | new_opt_data;
+
+            spi_frame_number++;
+        }
+    }
 
-    // Add new byte to received data
-    opt_spi_data = (opt_spi_data << 8) | new_byte;
-    // Received one more byte
-    opt_spi_num_bytes++;
+    // reset counter
+    spi_frame_number = 0;
 }
+
+
+// returns number of expected bytes to received from SPI, based on opcode cmd
+uint8_t opcode_to_num_bytes(uint8_t opcode){
+    switch (opcode){
+        case CMD_GET_READING:
+            return NUM_GET_READING;
+        default:
+            return 0;
+    }
+}
+
+
+// driver function to return 24-bit ADC data from wells
+// well_data = 
+uint32_t get_measurement(uint8_t well_data){
+    opt_spi_data = 0;       // clear previous data
+
+    uint8_t opcode = CMD_GET_READING;        // implied opcode is get reading
+    uint8_t num_expected_bytes = opcode_to_num_bytes(opcode);   // get number of bytes PAY is expectings
+
+    send_opt_spi_cmd(opcode, well_data);
+    receive_opt_data(num_expected_bytes);
+
+    return opt_spi_data;
```

Comment:
This is not how we want to get data because it's blocking for up to several seconds. Basically you want some "send command" function (`send_opt_spi_cmd()` as you correctly implemented above), and some "check if we received a response function". This function should check if the in progress flag is true and optical has asserted DATA_RDY, and if so it receives 3 bytes in a blocking sequence and adds a message to the CAN TX queue with the data. If the condition of in progress and DATA_RDY is false, it just returns and doesn't do anything.